client/macro_processor.py

```diff
@@ -0,0 +1,141 @@
+    Sql Translation job.
+
+    """
+    def __init__(self, context: Namespace):
+        self.context = context
+        if context.macro_map:
+            self.enable_macro_substitution = True
+            self.macros_subst = MapBasedExpander(context.macro_map)
+        else:
+            self.enable_macro_substitution = False
+
+    def pre_process(self, input_dir: str, output_dir: str):
+        assert self.enable_macro_substitution, "Macros processing not enabled."
+        self.process_inputs(input_dir, output_dir)
+
+    def post_process(self, input_dir: str, output_dir: str):
+        assert self.enable_macro_substitution, "Macros processing not enabled."
+        self.process_inputs(input_dir, output_dir, use_reversed_map=True)
+
+    def process_inputs(self, input_dir: str, output_dir: str, use_reversed_map=False):
+        """Replaces macros with values for every file in the input folder and save outputs in a new folder.  Macro
+        replacement doesn't apply for files with extension of .zip, .json, and .csv.
+
+        Args:
+            input_dir: path to the input directory.
+            output_dir: path to the output directory.
+            use_reversed_map: whether to use the reversed macro_replacement_maps.
+        """
+        for file in os.listdir(input_dir):
+            file_path = join(input_dir, file)
+            if isfile(file_path):
+                if not file.lower().endswith(('.zip', '.json', '.csv')) and not file.startswith("."):
+                    print("Applying token maps to file %s" % file_path)
+                    self.process_file(file_path, input_dir, output_dir, use_reversed_map)
+                elif not file.startswith("."):
+                    shutil.copy(file_path, file_path.replace(input_dir, output_dir))
+
+    def process_file(self, file_path: str, input_dir: str, output_dir: str, use_reversed_map):
+        """Applies the macro maps to a specific file and saves the output file in the output directory with the same
+        subdirectory and name as the input file.
+        """
+        assert isfile(file_path), "Can't find a file at \"%s\"." % file_path
+
+        # Prepare the output file path.
+        output_file_path = file_path.replace(input_dir, output_dir)
+        # Creates a new output directory if not exists.
+        os.makedirs(dirname(output_file_path), exist_ok=True)
+        # Creates a write_stream.
+        write_stream = open(output_file_path, "w")
+        # Prepares the available regex patterns for this file.
+        reg_pattern_map, patterns = self.get_all_regex_pattern_mapping(file_path, input_dir, use_reversed_map)
+        with open(file_path) as f:
+            for line in f:
+                line = self.process_text(line, reg_pattern_map, patterns)
+                write_stream.write(line)
+        write_stream.close()
+
+    def process_text(self, input_text, reg_pattern_map, patterns):
```

Comment:
The idea of this method is to make it take only self and input_text, so that it can be overridden.

def preprocess_text(self, input_text):
and
def postprocess_text(self, output_text):

The refactoring done here is good as a matter of software engineering, but it's not quite in line with the CUJ for modifiability.

Reply:
Updated per our earlier discussion.
Please take a look. Thanks